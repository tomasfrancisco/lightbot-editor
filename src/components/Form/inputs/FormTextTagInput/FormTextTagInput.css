.public-DraftStyleDefault-block {
  overflow-x: auto;
  white-space: pre;
}
