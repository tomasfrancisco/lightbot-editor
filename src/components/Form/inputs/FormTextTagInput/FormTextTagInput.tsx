import "./FormTextTagInput.css";

import { Tag as TagAntd } from "antd";
import { FormComponentProps } from "antd/lib/form";
import { GetFieldDecoratorOptions } from "antd/lib/form/Form";
import {
  CompositeDecorator,
  ContentState,
  DraftHandleValue,
  Editor,
  EditorState
} from "draft-js";
import * as React from "react";
import styled from "react-emotion";

const StyledTag = styled(TagAntd)`
  margin-right: 0px !important;
`;

const Tag = ({ children, color }) => (
  <StyledTag color={color}>{children}</StyledTag>
);

export type FormTextTagInputProps = {
  value?: string;
  validTags: string[];
  onChange?(value: string): void;
};

type FormTextTagInputState = {
  editorState: EditorState;
};

export class FormTextTagInput extends React.Component<
  FormTextTagInputProps,
  FormTextTagInputState
> {
  constructor(props: FormTextTagInputProps) {
    super(props);

    const compositeDecorator = new CompositeDecorator([
      {
        strategy: this.keywordStrategy,
        component: Tag,
        props: {
          color: "green"
        }
      },
      {
        strategy: this.entityKeywordStrategy,
        component: Tag,
        props: {
          color: "blue"
        }
      },
      {
        strategy: this.notValidKeywordsStrategy,
        component: Tag,
        props: {
          color: "red"
        }
      }
    ]);

    this.state = {
      editorState: EditorState.createWithContent(
        ContentState.createFromText(props.value || ""),
        compositeDecorator
      )
    };
  }

  public render() {
    const { editorState } = this.state;

    return (
      <div className="ant-input">
        <Editor
          editorState={editorState}
          onChange={this.onChange}
          handleReturn={this.handleReturn}
        />
      </div>
    );
  }

  private handleReturn = (): DraftHandleValue => {
    return "handled";
  };

  private onChange = (editorState: EditorState) => {
    const { onChange } = this.props;

    this.setState({ editorState });

    if (onChange) {
      onChange(editorState.getCurrentContent().getPlainText());
    }
  };

  private entityKeywordStrategy = (contentBlock, callback, contentState) => {
    const regex = /(?<=\s|^)\$SYS[^ <$]*/gim;

    this.findWithRegex(regex, contentBlock, callback, false);
  };

  private notValidKeywordsStrategy = (contentBlock, callback, contentState) => {
    this.keywordStrategy(contentBlock, callback, contentState, false);
  };

  private keywordStrategy = (contentBlock, callback, _, includes = true) => {
    const regex = /(?<=\s|^)\$[^ <$]*/gm;

    this.findWithRegex(regex, contentBlock, callback, includes);
  };

  private findWithRegex = (regex, contentBlock, callback, includes) => {
    const { validTags } = this.props;
    const text = contentBlock.getText();
    let matchArr;
    let start;
    // tslint:disable-next-line
    while ((matchArr = regex.exec(text)) !== null) {
      start = matchArr.index;
      if (includes) {
        if (
          validTags.includes(text.substring(start, start + matchArr[0].length))
        ) {
          callback(start, start + matchArr[0].length);
        }
      } else {
        callback(start, start + matchArr[0].length);
      }
    }
  };
}

const getFieldValidatorKey = (itemKey: string, itemGroup?: string) => {
  if (itemGroup) {
    return `${itemGroup}[${itemKey}]`;
  } else {
    return itemKey;
  }
};

export type RenderFormTextTagInput = FormComponentProps &
  GetFieldDecoratorOptions & {
    itemKey: string;
    itemGroup?: string;
    value: any;
    validTags: string[];
    type?: string;
    errorMessage: string;
    getValueProps: any; // TODO: Remove once antd gets updated
    validator?(
      rule: any,
      value: any,
      callback: any,
      source?: any,
      options?: any
    ): any;
  };

export const renderFormTextTagInput = ({
  form,
  itemKey,
  itemGroup,
  value,
  errorMessage,
  getValueFromEvent,
  getValueProps,
  type,
  validator,
  validTags
}: RenderFormTextTagInput) =>
  form &&
  form.getFieldDecorator(getFieldValidatorKey(itemKey, itemGroup), {
    getValueFromEvent,
    // @ts-ignore
    getValueProps,
    initialValue: value,
    rules: [
      {
        message: errorMessage,
        type,
        validator,
        required: true
      }
    ],
    validateTrigger: ["onChange", "onBlur"]
  })(<FormTextTagInput validTags={validTags} />);
