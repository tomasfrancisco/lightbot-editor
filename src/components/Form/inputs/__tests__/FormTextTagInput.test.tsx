// import generateRandomKey from "draft-js/lib/generateRandomKey";
import { shallow } from "enzyme";
import * as React from "react";

import { FormTextTagInput } from "../FormTextTagInput";

// This mock prevents Draft.js of generating random keys and makes
// the snapshots consistent between runs
jest.mock("draft-js/lib/generateRandomKey", () => () => "testing-key");

describe("components/Form/inputs/FormTextTagInput", () => {
  let component;

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  beforeEach(() => {
    const validTags = ["$Hello", "$world"];
    const value = "$Hello $NotValidWord and plain text.";

    component = shallow(
      <FormTextTagInput validTags={validTags} value={value} />
    );
  });

  test("renders", () => {
    jest.runAllTimers();
    expect(component).toMatchSnapshot();
  });
});
