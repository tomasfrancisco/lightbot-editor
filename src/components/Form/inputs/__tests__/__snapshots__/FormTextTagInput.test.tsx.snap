// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Form/inputs/FormTextTagInput renders 1`] = `
<div
  className="ant-input"
>
  <DraftEditor
    blockRenderMap={
      Immutable.Map {
        "ordered-list-item": Object {
          "element": "li",
          "wrapper": <ol
            className="public-DraftStyleDefault-ol"
          />,
        },
        "header-six": Object {
          "element": "h6",
        },
        "header-four": Object {
          "element": "h4",
        },
        "header-one": Object {
          "element": "h1",
        },
        "unordered-list-item": Object {
          "element": "li",
          "wrapper": <ul
            className="public-DraftStyleDefault-ul"
          />,
        },
        "atomic": Object {
          "element": "figure",
        },
        "unstyled": Object {
          "aliasedElements": Array [
            "p",
          ],
          "element": "div",
        },
        "header-two": Object {
          "element": "h2",
        },
        "code-block": Object {
          "element": "pre",
          "wrapper": <pre
            className="public-DraftStyleDefault-pre"
          />,
        },
        "blockquote": Object {
          "element": "blockquote",
        },
        "header-five": Object {
          "element": "h5",
        },
        "header-three": Object {
          "element": "h3",
        },
      }
    }
    blockRendererFn={[Function]}
    blockStyleFn={[Function]}
    editorState={
      EditorState {
        "_immutable": Immutable.Record {
          "allowUndo": true,
          "currentContent": Immutable.Record {
            "entityMap": Object {
              "__add": [Function],
              "__create": [Function],
              "__get": [Function],
              "__getLastCreatedEntityKey": [Function],
              "__mergeData": [Function],
              "__replaceData": [Function],
              "add": [Function],
              "create": [Function],
              "get": [Function],
              "getLastCreatedEntityKey": [Function],
              "mergeData": [Function],
              "replaceData": [Function],
            },
            "blockMap": Immutable.OrderedMap {
              "testing-key": Immutable.Record {
                "key": "testing-key",
                "type": "unstyled",
                "text": "$Hello $NotValidWord and plain text.",
                "characterList": Immutable.List [
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                  Immutable.Record {
                    "style": Immutable.OrderedSet [],
                    "entity": null,
                  },
                ],
                "depth": 0,
                "data": Immutable.Map {},
              },
            },
            "selectionBefore": Immutable.Record {
              "anchorKey": "testing-key",
              "anchorOffset": 0,
              "focusKey": "testing-key",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "selectionAfter": Immutable.Record {
              "anchorKey": "testing-key",
              "anchorOffset": 0,
              "focusKey": "testing-key",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
          },
          "decorator": CompositeDraftDecorator {
            "_decorators": Array [
              Object {
                "component": [Function],
                "props": Object {
                  "color": "green",
                },
                "strategy": [Function],
              },
              Object {
                "component": [Function],
                "props": Object {
                  "color": "blue",
                },
                "strategy": [Function],
              },
              Object {
                "component": [Function],
                "props": Object {
                  "color": "red",
                },
                "strategy": [Function],
              },
            ],
          },
          "directionMap": Immutable.OrderedMap {
            "testing-key": "LTR",
          },
          "forceSelection": false,
          "inCompositionMode": false,
          "inlineStyleOverride": null,
          "lastChangeType": null,
          "nativelyRenderedContent": null,
          "redoStack": Immutable.Stack [],
          "selection": Immutable.Record {
            "anchorKey": "testing-key",
            "anchorOffset": 0,
            "focusKey": "testing-key",
            "focusOffset": 0,
            "isBackward": false,
            "hasFocus": false,
          },
          "treeMap": Immutable.OrderedMap {
            "testing-key": Immutable.List [
              Immutable.Record {
                "start": 0,
                "end": 6,
                "decoratorKey": "0.0",
                "leaves": Immutable.List [
                  Immutable.Record {
                    "start": 0,
                    "end": 6,
                  },
                ],
              },
              Immutable.Record {
                "start": 6,
                "end": 7,
                "decoratorKey": null,
                "leaves": Immutable.List [
                  Immutable.Record {
                    "start": 6,
                    "end": 7,
                  },
                ],
              },
              Immutable.Record {
                "start": 7,
                "end": 20,
                "decoratorKey": "2.0",
                "leaves": Immutable.List [
                  Immutable.Record {
                    "start": 7,
                    "end": 20,
                  },
                ],
              },
              Immutable.Record {
                "start": 20,
                "end": 36,
                "decoratorKey": null,
                "leaves": Immutable.List [
                  Immutable.Record {
                    "start": 20,
                    "end": 36,
                  },
                ],
              },
            ],
          },
          "undoStack": Immutable.Stack [],
        },
      }
    }
    handleReturn={[Function]}
    keyBindingFn={[Function]}
    onChange={[Function]}
    readOnly={false}
    spellCheck={false}
    stripPastedStyles={false}
  />
</div>
`;
